File: lib/src/selectors.dart
part of 'polars.dart';

/// The time-zone set used by a datetime datatype selector.
enum TimeZoneSelectorMode {
  any('any'),
  anySet('anySet'),
  unset('unset'),
  anyOf('anyOf'),
  unsetOrAnyOf('unsetOrAnyOf');

  const TimeZoneSelectorMode(this.wireName);
  final String wireName;
}

/// A schema-dependent column selector.
final class Selector {
  Selector._(this._runtime, int handle)
    : _owner = _HandleOwner(_runtime, handle) {
    _owner.attach();
  }

  final Polars _runtime;
  final _HandleOwner _owner;
  bool get isClosed => _owner.isClosed;
  void close() => _owner.close();

  Selector _binary(String operation, Selector other) {
    _runtime._requireSelector(other);
    final leases = _leaseAll([_owner, other._owner]);
    try {
      return _runtime._adoptSelector(
        _runtime._client.invokeSync('selectorBinary', {
          'left': leases[0].handle.toString(),
          'right': leases[1].handle.toString(),
          'op': operation,
        }),
      );
    } finally {
      _releaseAll(leases);
    }
  }

  Selector operator |(Selector other) => _binary('union', other);
  Selector operator &(Selector other) => _binary('intersection', other);
  Selector operator ^(Selector other) => _binary('symmetricDifference', other);
  Selector operator -(Selector other) => _binary('difference', other);

  /// Returns the complement of this selector.
  Selector operator ~() {
    final lease = _owner.lease('Selector');
    try {
      return _runtime._adoptSelector(
        _runtime._client.invokeSync('selectorNot', {
          'input': lease.handle.toString(),
        }),
      );
    } finally {
      lease.end();
    }
  }

  /// Converts this selector to a projection expression.
  Expr asExpr() {
    final lease = _owner.lease('Selector');
    try {
      return _runtime._adoptExpr(
        _runtime._client.invokeSync('selectorAsExpr', {
          'input': lease.handle.toString(),
        }),
      );
    } finally {
      lease.end();
    }
  }
}

/// A composable predicate over Polars datatypes.
final class DTypeSelector {
  DTypeSelector._(this._runtime, int handle)
    : _owner = _HandleOwner(_runtime, handle) {
    _owner.attach();
  }

  final Polars _runtime;
  final _HandleOwner _owner;
  bool get isClosed => _owner.isClosed;
  void close() => _owner.close();

  DTypeSelector _binary(String operation, DTypeSelector other) {
    _runtime._requireDTypeSelector(other);
    final leases = _leaseAll([_owner, other._owner]);
    try {
      return _runtime._adoptDTypeSelector(
        _runtime._client.invokeSync('dtypeSelectorBinary', {
          'left': leases[0].handle.toString(),
          'right': leases[1].handle.toString(),
          'op': operation,
        }),
      );
    } finally {
      _releaseAll(leases);
    }
  }

  DTypeSelector operator |(DTypeSelector other) => _binary('union', other);
  DTypeSelector operator &(DTypeSelector other) =>
      _binary('intersection', other);
  DTypeSelector operator ^(DTypeSelector other) =>
      _binary('symmetricDifference', other);
  DTypeSelector operator -(DTypeSelector other) => _binary('difference', other);

  DTypeSelector operator ~() {
    final lease = _owner.lease('DTypeSelector');
    try {
      return _runtime._adoptDTypeSelector(
        _runtime._client.invokeSync('dtypeSelectorNot', {
          'input': lease.handle.toString(),
        }),
      );
    } finally {
      lease.end();
    }
  }

  Selector asSelector() {
    final lease = _owner.lease('DTypeSelector');
    try {
      return _runtime._adoptSelector(
        _runtime._client.invokeSync('dtypeSelectorAsSelector', {
          'input': lease.handle.toString(),
        }),
      );
    } finally {
      lease.end();
    }
  }

  bool matches(DType dtype) {
    final lease = _owner.lease('DTypeSelector');
    try {
      return _runtime._client.invokeSync('dtypeSelectorMatches', {
            'input': lease.handle.toString(),
            'dtype': dtype.toJson(),
          })['matches']
          as bool;
    } finally {
      lease.end();
    }
  }
}

/// Runtime-bound constructors for selectors and datatype selectors.
final class SelectorFactory {
  const SelectorFactory._(this._runtime);
  final Polars _runtime;

  Selector _selector(
    String command, [
    Map<String, Object?> fields = const {},
  ]) => _runtime._adoptSelector(_runtime._client.invokeSync(command, fields));

  DTypeSelector _dtype(String kind, [Map<String, Object?> fields = const {}]) =>
      _runtime._adoptDTypeSelector(
        _runtime._client.invokeSync('dtypeSelectorCreate', {
          'kind': kind,
          ...fields,
        }),
      );

  Selector all() => _selector('selectorAll');
  Selector empty() => _selector('selectorEmpty');

  Selector byName(
    Iterable<String> names, {
    bool strict = true,
    bool expandPatterns = false,
  }) {
    final values = _selectorNames(names, 'names');
    return _selector('selectorByName', {
      'names': values,
      'strict': strict,
      'expandPatterns': expandPatterns,
    });
  }

  Selector byIndex(Iterable<int> indices, {bool strict = true}) {
    final values = List<int>.unmodifiable(indices);
    if (values.isEmpty || values.length > 10000) {
      throw ArgumentError.value(
        indices,
        'indices',
        'must contain 1 to 10000 values',
      );
    }
    for (final value in values) {
      if (value < -0x8000000000000000 || value > 0x7fffffffffffffff) {
        throw RangeError.value(value, 'indices', 'must fit signed 64-bit');
      }
    }
    return _selector('selectorByIndex', {'indices': values, 'strict': strict});
  }

  Selector matches(String pattern) {
    _validateName(pattern, 'pattern');
    return _selector('selectorMatches', {'pattern': pattern});
  }

  Selector byDType(DTypeSelector selector) {
    _runtime._requireDTypeSelector(selector);
    return selector.asSelector();
  }

  /// Selects columns whose datatype is one of [dtypes].
  Selector dtypes(Iterable<DType> dtypes) => anyOf(dtypes)._intoSelector();

  DTypeSelector dtypeAll() => _dtype('all');
  DTypeSelector dtypeEmpty() => _dtype('empty');
  DTypeSelector anyOf(Iterable<DType> dtypes) {
    final values = List<DType>.unmodifiable(dtypes);
    if (values.isEmpty || values.length > 10000) {
      throw ArgumentError.value(
        dtypes,
        'dtypes',
        'must contain 1 to 10000 values',
      );
    }
    return _dtype('anyOf', {
      'dtypes': values.map((value) => value.toJson()).toList(growable: false),
    });
  }

  DTypeSelector integerDTypes() => _dtype('integer');
  DTypeSelector unsignedIntegerDTypes() => _dtype('unsignedInteger');
  DTypeSelector signedIntegerDTypes() => _dtype('signedInteger');
  DTypeSelector floatingDTypes() => _dtype('floating');
  DTypeSelector enumDTypes() => _dtype('enum');
  DTypeSelector categoricalDTypes() => _dtype('categorical');
  DTypeSelector nestedDTypes() => _dtype('nested');
  DTypeSelector structDTypes() => _dtype('struct');
  DTypeSelector decimalDTypes() => _dtype('decimal');
  DTypeSelector numericDTypes() => _dtype('numeric');
  DTypeSelector temporalDTypes() => _dtype('temporal');
  DTypeSelector objectDTypes() => _dtype('object');

  DTypeSelector listDTypes({DTypeSelector? inner}) =>
      _nestedDType('list', inner);

  DTypeSelector arrayDTypes({DTypeSelector? inner, int? width}) {
    if (width != null && width <= 0) {
      throw RangeError.value(width, 'width', 'must be positive');
    }
    return _nestedDType('array', inner, {if (width != null) 'width': width});
  }

  DTypeSelector _nestedDType(
    String kind,
    DTypeSelector? inner, [
    Map<String, Object?> fields = const {},
  ]) {
    if (inner != null) _runtime._requireDTypeSelector(inner);
    final lease = inner?._owner.lease('DTypeSelector');
    try {
      return _dtype(kind, {
        if (lease != null) 'inner': lease.handle.toString(),
        ...fields,
      });
    } finally {
      lease?.end();
    }
  }

  DTypeSelector datetimeDTypes({
    Iterable<TimeUnit> units = TimeUnit.values,
    TimeZoneSelectorMode timeZoneMode = TimeZoneSelectorMode.any,
    Iterable<String> timeZones = const [],
  }) {
    final unitValues = List<TimeUnit>.unmodifiable(units);
    if (unitValues.isEmpty) {
      throw ArgumentError.value(units, 'units', 'must not be empty');
    }
    final zones = List<String>.unmodifiable(timeZones);
    final needsZones =
        timeZoneMode == TimeZoneSelectorMode.anyOf ||
        timeZoneMode == TimeZoneSelectorMode.unsetOrAnyOf;
    if (needsZones != zones.isNotEmpty || zones.any((zone) => zone.isEmpty)) {
      throw ArgumentError.value(
        timeZones,
        'timeZones',
        'must be non-empty exactly for anyOf/unsetOrAnyOf',
      );
    }
    return _dtype('datetime', {
      'units': unitValues.map((unit) => unit.json).toList(growable: false),
      'timeZoneMode': timeZoneMode.wireName,
      'timeZones': zones,
    });
  }

  DTypeSelector durationDTypes({Iterable<TimeUnit> units = TimeUnit.values}) {
    final values = List<TimeUnit>.unmodifiable(units);
    if (values.isEmpty) {
      throw ArgumentError.value(units, 'units', 'must not be empty');
    }
    return _dtype('duration', {
      'units': values.map((unit) => unit.json).toList(growable: false),
    });
  }

  Selector integer() => integerDTypes()._intoSelector();
  Selector unsignedInteger() => unsignedIntegerDTypes()._intoSelector();
  Selector signedInteger() => signedIntegerDTypes()._intoSelector();
  Selector floating() => floatingDTypes()._intoSelector();
  Selector categorical() => categoricalDTypes()._intoSelector();
  Selector enumeration() => enumDTypes()._intoSelector();
  Selector nested() => nestedDTypes()._intoSelector();
  Selector structs() => structDTypes()._intoSelector();
  Selector decimals() => decimalDTypes()._intoSelector();
  Selector numeric() => numericDTypes()._intoSelector();
  Selector temporal() => temporalDTypes()._intoSelector();
  Selector objects() => objectDTypes()._intoSelector();
}

extension on DTypeSelector {
  Selector _intoSelector() {
    try {
      return asSelector();
    } finally {
      close();
    }
  }
}

List<String> _selectorNames(Iterable<String> names, String argument) {
  final values = List<String>.unmodifiable(names);
  if (values.isEmpty ||
      values.length > 10000 ||
      values.any((value) => value.isEmpty)) {
    throw ArgumentError.value(
      names,
      argument,
      'must contain 1 to 10000 non-empty values',
    );
  }
  return values;
}

File: native/polars_ffi/src/selectors.rs
use crate::{
    bindings as b, dtype,
    error::{EngineError, Result},
    registry::{self, Entry},
};
use polars::prelude::*;
use serde_json::{json, Value};
use std::sync::Arc;

fn response(handle: u64, kind: &str) -> Value {
    json!({"handle": handle.to_string(), "kind": kind})
}

fn insert_selector(value: Selector) -> Result<Value> {
    Ok(response(
        registry::insert(Entry::Selector(value))?,
        "selector",
    ))
}

fn insert_dtype_selector(value: DataTypeSelector) -> Result<Value> {
    Ok(response(
        registry::insert(Entry::DataTypeSelector(value))?,
        "dtypeSelector",
    ))
}

fn fields(v: &Value, extra: &[&str]) -> Result<()> {
    let mut allowed = vec!["protocol", "command"];
    allowed.extend(extra);
    b::validate_fields(v, &allowed)
}

fn strings(v: &Value, key: &str) -> Result<Vec<String>> {
    Ok(b::names(v, key)?
        .into_iter()
        .map(|name| name.to_string())
        .collect())
}

fn selector_binary(v: &Value) -> Result<Selector> {
    let left = registry::selector(b::handle(v, "left")?)?;
    let right = registry::selector(b::handle(v, "right")?)?;
    Ok(match b::string(v, "op")? {
        "union" => left | right,
        "intersection" => left & right,
        "symmetricDifference" => left ^ right,
        "difference" => left - right,
        op => {
            return Err(EngineError::Invalid(format!(
                "unknown selector operation '{op}'"
            )))
        }
    })
}

fn dtype_selector_binary(v: &Value) -> Result<DataTypeSelector> {
    let left = registry::dtype_selector(b::handle(v, "left")?)?;
    let right = registry::dtype_selector(b::handle(v, "right")?)?;
    Ok(match b::string(v, "op")? {
        "union" => left | right,
        "intersection" => left & right,
        "symmetricDifference" => left ^ right,
        "difference" => left - right,
        op => {
            return Err(EngineError::Invalid(format!(
                "unknown datatype selector operation '{op}'"
            )))
        }
    })
}

fn time_units(v: &Value) -> Result<TimeUnitSet> {
    let values = strings(v, "units")?;
    if values.is_empty() {
        return Err(EngineError::Invalid("'units' must not be empty".into()));
    }
    let mut units = TimeUnitSet::empty();
    for value in values {
        units |= match value.as_str() {
            "nanoseconds" => TimeUnitSet::NANO_SECONDS,
            "microseconds" => TimeUnitSet::MICRO_SECONDS,
            "milliseconds" => TimeUnitSet::MILLI_SECONDS,
            _ => return Err(EngineError::Invalid(format!("unknown time unit '{value}'"))),
        };
    }
    Ok(units)
}

fn time_zones(v: &Value) -> Result<TimeZoneSet> {
    let mode = b::string(v, "timeZoneMode")?;
    let zones = if v.get("timeZones").is_some() {
        strings(v, "timeZones")?
    } else {
        Vec::new()
    };
    let requires_zones = matches!(mode, "anyOf" | "unsetOrAnyOf");
    if requires_zones != !zones.is_empty() {
        return Err(EngineError::Invalid(
            "timeZones must be non-empty exactly for anyOf/unsetOrAnyOf".into(),
        ));
    }
    let zones = zones
        .into_iter()
        .map(|zone| {
            TimeZone::opt_try_new(Some(zone))?
                .ok_or_else(|| EngineError::Invalid("time zone must not be empty".into()))
        })
        .collect::<Result<Vec<_>>>()?;
    Ok(match mode {
        "any" => TimeZoneSet::Any,
        "anySet" => TimeZoneSet::AnySet,
        "unset" => TimeZoneSet::Unset,
        "anyOf" => TimeZoneSet::AnyOf(zones.into()),
        "unsetOrAnyOf" => TimeZoneSet::UnsetOrAnyOf(zones.into()),
        _ => {
            return Err(EngineError::Invalid(format!(
                "unknown time zone mode '{mode}'"
            )))
        }
    })
}

fn optional_dtype_selector(v: &Value) -> Result<Option<Arc<DataTypeSelector>>> {
    v.get("inner")
        .map(|_| registry::dtype_selector(b::handle(v, "inner")?).map(Arc::new))
        .transpose()
}

fn create_dtype_selector(v: &Value) -> Result<DataTypeSelector> {
    let kind = b::string(v, "kind")?;
    let allowed: &[&str] = match kind {
        "anyOf" => &["kind", "dtypes"],
        "list" => &["kind", "inner"],
        "array" => &["kind", "inner", "width"],
        "datetime" => &["kind", "units", "timeZoneMode", "timeZones"],
        "duration" => &["kind", "units"],
        _ => &["kind"],
    };
    fields(v, allowed)?;
    Ok(match kind {
        "all" => DataTypeSelector::Wildcard,
        "empty" => DataTypeSelector::Empty,
        "anyOf" => {
            let values = b::req(v, "dtypes")?
                .as_array()
                .ok_or_else(|| EngineError::Invalid("'dtypes' must be an array".into()))?;
            if values.is_empty() || values.len() > 10_000 {
                return Err(EngineError::Invalid(
                    "'dtypes' must contain 1 to 10000 values".into(),
                ));
            }
            DataTypeSelector::AnyOf(
                values
                    .iter()
                    .map(dtype::parse)
                    .collect::<Result<Vec<_>>>()?
                    .into(),
            )
        }
        "integer" => DataTypeSelector::Integer,
        "unsignedInteger" => DataTypeSelector::UnsignedInteger,
        "signedInteger" => DataTypeSelector::SignedInteger,
        "floating" => DataTypeSelector::Float,
        "enum" => DataTypeSelector::Enum,
        "categorical" => DataTypeSelector::Categorical,
        "nested" => DataTypeSelector::Nested,
        "list" => DataTypeSelector::List(optional_dtype_selector(v)?),
        "array" => {
            let width = v
                .get("width")
                .map(|value| {
                    value
                        .as_u64()
                        .filter(|width| *width > 0)
                        .ok_or_else(|| EngineError::Invalid("'width' must be positive".into()))
                        .and_then(|width| {
                            usize::try_from(width)
                                .map_err(|_| EngineError::Invalid("'width' is too large".into()))
                        })
                })
                .transpose()?;
            DataTypeSelector::Array(optional_dtype_selector(v)?, width)
        }
        "struct" => DataTypeSelector::Struct,
        "decimal" => DataTypeSelector::Decimal,
        "numeric" => DataTypeSelector::Numeric,
        "temporal" => DataTypeSelector::Temporal,
        "datetime" => DataTypeSelector::Datetime(time_units(v)?, time_zones(v)?),
        "duration" => DataTypeSelector::Duration(time_units(v)?),
        "object" => DataTypeSelector::Object,
        _ => {
            return Err(EngineError::Invalid(format!(
                "unknown datatype selector kind '{kind}'"
            )))
        }
    })
}

pub fn invoke(command: &str, v: &Value) -> Result<Value> {
    match command {
        "selectorAll" => {
            fields(v, &[])?;
            insert_selector(Selector::Wildcard)
        }
        "selectorEmpty" => {
            fields(v, &[])?;
            insert_selector(Selector::Empty)
        }
        "selectorByName" => {
            fields(v, &["names", "strict", "expandPatterns"])?;
            let names = b::names(v, "names")?;
            if names.is_empty() {
                return Err(EngineError::Invalid("'names' must not be empty".into()));
            }
            insert_selector(polars::lazy::dsl::by_name(
                names,
                b::optional_bool(v, "strict", true)?,
                b::optional_bool(v, "expandPatterns", false)?,
            ))
        }
        "selectorByIndex" => {
            fields(v, &["indices", "strict"])?;
            let values = b::req(v, "indices")?
                .as_array()
                .ok_or_else(|| EngineError::Invalid("'indices' must be an array".into()))?;
            if values.is_empty() || values.len() > 10_000 {
                return Err(EngineError::Invalid(
                    "'indices' must contain 1 to 10000 values".into(),
                ));
            }
            let indices = values
                .iter()
                .map(|value| {
                    value
                        .as_i64()
                        .ok_or_else(|| EngineError::Invalid("indices must be i64 values".into()))
                })
                .collect::<Result<Vec<_>>>()?;
            insert_selector(Selector::ByIndex {
                indices: indices.into(),
                strict: b::optional_bool(v, "strict", true)?,
            })
        }
        "selectorMatches" => {
            fields(v, &["pattern"])?;
            insert_selector(Selector::Matches(b::string(v, "pattern")?.into()))
        }
        "selectorBinary" => {
            fields(v, &["left", "right", "op"])?;
            insert_selector(selector_binary(v)?)
        }
        "selectorNot" => {
            fields(v, &["input"])?;
            insert_selector(!registry::selector(b::handle(v, "input")?)?)
        }
        "selectorAsExpr" => {
            fields(v, &["input"])?;
            let expr = registry::selector(b::handle(v, "input")?)?.as_expr();
            Ok(response(registry::insert(Entry::Expr(expr))?, "expr"))
        }
        "dtypeSelectorCreate" => insert_dtype_selector(create_dtype_selector(v)?),
        "dtypeSelectorBinary" => {
            fields(v, &["left", "right", "op"])?;
            insert_dtype_selector(dtype_selector_binary(v)?)
        }
        "dtypeSelectorNot" => {
            fields(v, &["input"])?;
            insert_dtype_selector(!registry::dtype_selector(b::handle(v, "input")?)?)
        }
        "dtypeSelectorAsSelector" => {
            fields(v, &["input"])?;
            insert_selector(registry::dtype_selector(b::handle(v, "input")?)?.as_selector())
        }
        "dtypeSelectorMatches" => {
            fields(v, &["input", "dtype"])?;
            let selector = registry::dtype_selector(b::handle(v, "input")?)?;
            Ok(json!({"matches": selector.matches(&dtype::parse(b::req(v, "dtype")?)?)}))
        }
        _ => Err(EngineError::Invalid(format!(
            "unknown selector command '{command}'"
        ))),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn selector_handles_branch_without_consuming_inputs() {
        let all = registry::insert(Entry::Selector(Selector::Wildcard)).unwrap();
        let numeric = registry::insert(Entry::DataTypeSelector(DataTypeSelector::Numeric)).unwrap();
        let numeric_selector = registry::dtype_selector(numeric).unwrap().as_selector();
        let numeric_selector = registry::insert(Entry::Selector(numeric_selector)).unwrap();
        let difference =
            registry::selector(all).unwrap() - registry::selector(numeric_selector).unwrap();
        assert!(matches!(difference, Selector::Difference(_, _)));
        assert!(registry::selector(all).is_ok());
        assert!(registry::dtype_selector(numeric).is_ok());
        for handle in [all, numeric, numeric_selector] {
            registry::release(handle).unwrap();
        }
    }
}
